Implement basictracer for Javascript

CatTail opened the pull request.

CatTail committed:
feat: implement basictracer based on opentracing API

bhs commented:
Can you do something like https://github.com/opentracing/opentracing-javascript/blob/master/.gitignore before we get started? Thanks.

bhs commented:
(fine to squash the original commit)

CatTail commented:
@bensigelman Just update `.gitignore`, is this what you mean?

bhs commented:
@CatTail yes; I was also suggesting that you squash your commits (via `git rebase -i` or similar) so the `dist/` files never end up in the repo, but that is not important.

bhs commented on src/tracer.js:
```diff
@@ -0,0 +1,130 @@
+        this._interface = inf
+    }
+    /**
+     * Starts and returns a new Span representing a logical unit of work.
```
nit: no need for the repeats of OpenTracing comments... it's probably better to just include a pointer in a `BasicTracer` class comment to the OpenTracing-js Tracer spec.

bhs commented on src/tracer.js:
```diff
@@ -0,0 +1,130 @@
+import Recorder from './recorder'
+
+export default class BasicTracer {
+    constructor({sampler, recorder} = {}) {
```
please do include comments for the constructor and the expected types of the sampler and recorder

bhs commented on src/tracer.js:
```diff
@@ -0,0 +1,130 @@
+    isSample(span, parent) {
+        return this._sampler.isSample(span, parent)
+    }
+    record(span) {
```
should this be private?

bhs commented on src/tracer.js:
```diff
@@ -0,0 +1,130 @@
+        return span
+    }
+
+    isSample(span, parent) {
```
why is the parent needed?

bhs replied:
also: should this be private / `_isSampled`?

bhs commented on src/state.proto:
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
+
+message TracerState {
```
this really should be brought in through a submodule or something, but let's leave it this way for now.

bhs commented on src/index.js:
```diff
@@ -0,0 +1,10 @@
+import Backend from './tracer'
+
+tracer.configure = (options) => {
+    tracer.initGlobalTracer(new Backend(options))
```
Sorry, where is Backend defined?

CatTail replied:
src/tracer.js is the `Backend`, and the name coming from https://github.com/opentracing/opentracing-javascript/blob/master/README.md example

``` js
var Tracer = require('opentracing');
var TracingBackend = require('tracing-implementation-of-your-choice');

Tracer.initGlobalTracer(TracingBackend.create());
```

CatTail commented:
@bensigelman 

> why is the parent needed?

In code

``` js
isSample(span, parent) {
}
```

Sometimes we need to decide whether current span should be used based on parent(carrier) information.

bhs commented on src/sampler.js:
```diff
@@ -0,0 +1,8 @@
+export default class DefaultSampler {
+    isSample(span, parent) {
```
per the previous discussion (which is lost due to the squash): IMO it's not necessary to consider whether a non-root span (i.e., a span with a parent) should be sampled in BasicTracer... every tracing system I'm aware of just inherits the parent's sampling status in that case.

Also, `s/isSample/isSampled/`... and please add a brief comment explaining the semantics and expected parameter and return types.

bhs commented on src/recorder.js:
```diff
@@ -0,0 +1,17 @@
+export default class Recorder {
```
Please add a good comment for `Recorder` and `record`, as this is the most important public concept in `basictracer`.

bhs commented on src/recorder.js:
```diff
@@ -0,0 +1,17 @@
+export default class Recorder {
+    record(span) {
+        console.log(JSON.stringify({ // eslint-disable-line
```
something like a `DebuggingRecorder` would be great, but the default Recorder should do nothing at all.

bhs commented on src/span.js:
```diff
@@ -0,0 +1,81 @@
+        this._tracer._record(this)
+    }
+
+    static generateUUID() {
```
I'm trusting you that this is the right way to generate a random int64 in javascript :)

bhs commented on src/propagation.js:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    inject(span, carrier) {
+        carrier[FIELD_NAME_TRACE_ID] = span.traceId
```
it's important for this to do an _unsigned_ int64->string conversion (same with below, and with the other direction in join).

bhs commented on src/__tests__/tracker-test.js:
```diff
@@ -0,0 +1,76 @@
+        should(span.baggage).eql(parentSpan.baggage)
+    })
+
+    it('should join binary span', () => {
```
thanks for these tests, btw.

bhs commented on Readme.md:
```diff
@@ -0,0 +1,56 @@
+# basictracer-javascript [![NPM version][npm-image]][npm-url] [![build status][travis-image]][travis-url] [![Test coverage][coveralls-image]][coveralls-url]
```
I have a bunch of thoughts about the README, but as a thank-you to you for doing all of this work, I can just make README changes in my own followup PR :)

CatTail committed:
fix: text format inject and join don't convert Long type

Also change signature of Sampler and others, see
https://github.com/opentracing/basictracer-javascript/pull/1.

CatTail commented:
@bensigelman Just push a new commit, any problem?

bhs commented on src/sampler.js:
```diff
@@ -0,0 +1,9 @@
+export class DefaultSampler {
+    /**
+     * @param  {Span} span Span to be determined.
```
BasicSpan or opentracing Span?

bhs commented on src/tracer.js:
```diff
@@ -0,0 +1,73 @@
+
+    startSpan(fields) {
+        // Interface or Implementation argument
+        // https://github.com/opentracing/opentracing-javascript/pull/29
```
it didn't look like https://github.com/opentracing/opentracing-javascript/pull/29 was going to go through... isn't it safe to assume that `fields.parent` is an opentracing `Span` and thus has an `imp()` method?

CatTail replied:
Maybe, but opentracing library don't expose it's `Tracer` class, so there is no way to do that

bcronin replied:
I'm a little confused as well. I would have thought that `fields.parent` is _always_ a valid OpenTracing Span "Interface" object in the call to `startSpan`. Therefore, line 44 could simply be `parent: fields.parent.imp()` and line 39-41 could be removed. I'm not sure I understand how `Tracer` is involved.

My apologies if I'm missing something obvious here!

CatTail replied:
You are right, the only case is in [unittest](https://github.com/CatTail/basictracer-javascript-1/blob/4a50f7b1d163126df6b5bd275d662137f0d1c0a1/src/__tests__/tracker-test.js#L68). I'm going to change the unittest.

bhs commented:
Basically LGTM! Thanks. I have some nits about comments and documentation but it makes more sense for me to just suggest those changes separately, I think.

We should resolve the question about https://github.com/opentracing/opentracing-javascript/pull/29 and merge. Thanks again.

CatTail committed:
docs: explain what type recorder will receive

CatTail commented:
@bensigelman Just add a little comment for `Recorder`, but I'm not a native english speaker, my comment maybe not appropriate (accurate)

CatTail committed:
feat: support http header propagation

CatTail committed:
0.2.0

CatTail committed:
fix: represent fixed64 as decimal

CatTail committed:
0.2.3

bhs commented:
@CatTail thanks for the updates. I'm curious to hear your response to @bcronin's most recent comment (about `imp()`).

bcronin commented:
This looks like a lot of good code. Thanks @CatTail .

One issue / question that can possibly be addressed _after_ this pull request:

It looks like this code is currently Node-specific (for example, it uses `Buffer`).  How difficult would it be to make this code work both on the browser and in Node?  One option would be to rename this repository `basictracer-node` and keep it Node-specific. Another option would be to make it "universal" so that it works on both the browser and Node.  I personally don't have a strong opinion on which approach is better.  In either case, I think this is a good starting point and browser-compatibility could be addressed in future changes.

CatTail commented:
@bcronin There are work need to be done to use in browser side. How about figure it out in future version?

CatTail commented:
@bcronin @bensigelman I think the bridge pattern limit the implementation's ability to extend. For example, how can I add a format support HTTP format?

Text format is not suitable because add `x-` prefix for HTTP header is more appropriate. Checkout [my refactor version of basictracer](https://github.com/CatTail/basictracer-javascript/blob/0a73ec2a190268a4b54aa26c9e05bb9b93476173/src/propagation.js#L49), it directly expose tracer rather use bridge pattern.

CatTail committed:
style: parent always being interface

bhs commented:
@CatTail FYI, the `X-` prefix has been deprecated since 2012: https://tools.ietf.org/html/rfc6648

Regardless, though, the idea with the "carriers" is that they can be adapted to HTTP headers or to other key:value maps (for example, grpc.io "metadata": https://github.com/grpc/grpc/blob/release-0_13/src/node/src/metadata.js) without requiring any additional work from Tracer implementations.

bhs commented on src/tracer.js:
```diff
@@ -0,0 +1,74 @@
+     * Constructs basic tracer instance with provided Sampler and Recorder
+     * @param  {Object} options
+     *         Optional associative array of fields.
+     *         - `sampler` {Sampler} Optional object with `isSample` method, the
```
nit: "isSampled" in the comment

bhs commented on src/index.js:
```diff
@@ -0,0 +1,20 @@
+
+// Little hack because interface don't expose a way to extend supported
+// carrier format
+tracer.FORMAT_HTTP_HEADER = 'http_header'
```
I don't understand why this is here... can you explain?

CatTail replied:
Because `opentracing-javascript` don't have a way to add additional format

bcronin commented:
@CatTail regarding the refactor that "directly expose tracer rather use bridge pattern.":

My concern here (and I'm glad to be corrected if I am missing something) is this removes what seems to me to be a necessary level of indirection. It would be great if third-party packages (for example, express.js) could be instrumented with OpenTracing _but not tied to an implementation_. That means that express.js package code needs to refer to an interface only.  If I then write my web application, let's call it MyApp, that _uses_ express.js, I need a way to connect express.js' instrumented code to the _implementation_ that MyApp is using.

This could be done with true global variables (using the `global` built-in variable) but there are problems with that (namely, many JS developers would consider it a bad practice and unit testing tools like `mocha` will complain by default about this). It could also be done by adding the tracer implementation to the initialization of express.js, but that would mean every single OpenTracing-compatible package needs to modify their initialization API and each one needs to be passed that implementation object  (and continue to pass it to its dependencies in turn); that is certainly _possible_, but is inconvenient and error-prone, and also quite awkward in utility libraries that don't have a specific initialize call in their API.

If there's a good way to instrument a third-party package with just the interface and then, in an uncomplicated way, connect (at runtime) a package to the host application's tracing implementation, I would enjoy hearing it and talking about alternatives to the bridge pattern.

CatTail commented:
@bcronin I don't have any good idea to restrict implementation API. Compared to [basictracer-python](https://github.com/opentracing/basictracer-python), they use inheritance to defined API and extend the ability to allow implementation to do more, such as support more carrier format.

bcronin commented:
@CatTail , I think I partly understand what you are saying. My understanding is that `FORMAT_HTTP_HEADER` is _not_ part of the OpenTracing standard; therefore, I'd argue it _should not_ be part of the `opentracing` package / namespace. From that perspective, I see it as potentially a good thing that a custom format can't be added to the set of constants exposed by the `opentracing` package. 

That said, the format argument to `join` and `inject` is defined as a `string`. Nothing prevents the implementation package from defining a new constant and using that. I don't fully understand how the current code prevents more carrier formats from being defined. They won't be under the same package / namespace, but again, I think that is correct since they are (by definition) not part of the standard.

Apologies if I am misunderstanding what you are saying!

bhs commented:
I think @bcronin is right about the supported formats... I think we should just merge this and make changes as followups. It's already in a useful place and is much better than what's at `master` (which is still empty).

I'll create some issues to track things that still need to be done.

bhs commented:
Issues opened: #2 #3 #4
